applications/FluidDynamicsApplication/custom_elements/data_containers/qs_vms/qs_vms_adjoint_element_data.h

```diff
@@ -0,0 +1,135 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_QS_VMS_ADJOINT_ELEMENT_DATA_H)
+#define KRATOS_QS_VMS_ADJOINT_ELEMENT_DATA_H
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/node.h"
+#include "includes/process_info.h"
+#include "geometries/geometry.h"
+#include "geometries/geometry_data.h"
+
+// Application includes
+#include "custom_elements/data_containers/qs_vms/qs_vms_derivative_utilities.h"
+#include "custom_elements/data_containers/qs_vms/qs_vms_residual_derivatives.h"
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+template <unsigned int TDim, unsigned int TNumNodes>
+class QSVMSAdjointElementData
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    using NodeType = Node<3>;
+
+    using GeometryType = Geometry<NodeType>;
+
+    using TResidualsDerivatives = QSVMSResidualDerivatives<TDim, TNumNodes>;
+
+    ///@}
+    ///@name Static Operations
+    ///@{
+
+    static void Check(
+        const Element& rElement,
+        const ProcessInfo& rProcessInfo);
+
+    static std::vector<const Variable<double>*> GetDofVariablesList();
+
+    ///@}
+    ///@name Classes
+    ///@{
+
+    class Primal
+    {
+    public:
+        ///@name Type Definitions
+        ///@{
+
+        using Data = typename TResidualsDerivatives::Data;
+
+        using ResidualsContributions = typename TResidualsDerivatives::ResidualsContributions;
+
+        ///@}
+    };
+
+    class StateDerivatives
+    {
+    public:
+        ///@name Type Definitions
+        ///@{
+
+        class SecondDerivatives
```

Comment:
I think that 3 levels of internal classes is way cumbersome... I'd use `StateSecondDerivatives` and `StateFirstDerivatives`. Not sure if this is only a detail or affects the entire framework, Just suggesting.

Reply:
Done.... Reduced to flat hierrachy, and got rid of middle level classes as well. Now everything is more simpler and elegant with fold expressions in C++17. 👍